### ISSUE TITLE
Error, all predictors are trivial

### ISSUE THREAD

**Issue by YangBen0528** [dougspeed/LDAK#11]
Hi:
I used custom annotation information to generate matrix and tagging file using the LDAK-Thin model, and used LDAK-Thin to generate Per-predictor heritabilities. However, when I calculated the effect value of each SNP using --elastic, this error occurred：
**Error, all predictors are trivial.**
When I use the BLD model, the same error occurs. Do you have any good solutions?

**Comment by YangBen0528** [comment]
Thank you for your attention to my previous inquiry! I am writing to provide a detailed description of the issue I encountered while using LDAK, and I would greatly appreciate your guidance.I am working on implementing PRS prediction in a non-human species. My goal is to utilize custom annotation files to improve the predictive accuracy for traits and explore the enrichment of different annotations. While similar studies have been extensively conducted in humans, there is a lack of reports in the species I am studying.

The steps I followed are as follows:（1）I generated LDAK-Thin.tagging and LDAK-Thin.matrix using the annotation files with the command: 
ldak6 --max-threads 100 --calc-tagging LDAK-Thin --bfile geno --power .25 --annotation-number 37 --annotation-prefix ./bld/bld

（2）Using the files generated in the first step, I estimated per-predictor heritabilities:ldak6 --sum-hers thin.ldak --tagfile LDAK-Thin.tagging --summary time.summary --matrix LDAK-Thin.matrix --max-threads 20
（3）Finally, I calculated SNP effect sizes based on per-predictor heritabilities to predict PRS for another population. Since I have individual-level data, I utilized the Elastic Predict method.
 ldak6 --elastic elastic  --pheno ./pheno.txt  --bfile /public1/node3_yangb/imputation/beagel/geno --mpheno 1 --LOCO NO --ind-hers thin.ldak.ind.hers

However, when I ran the third part, I encountered the following error: Error, all predictors are trivial
Detailed Error Log:
Reading IDs for 1138 samples from /public1/node3_yangb/imputation/beagel/geno.fam
Checking responses for 1138 samples from ./pheno.txt
Reading details for 9226528 predictors from /public1/node3_yangb/imputation/beagel/geno.bim
Data contain 1138 samples and 9226528 predictors
Reading heritabilities for 9226528 predictors from thin.ldak.ind.hers
By ignoring those with non-positive heritability, number of predictors reduced to 3560134
Sum of per-predictor heritabilities: 2.393951
Warning, this must be less than one; will rescale pre-predictor heritabilities so that they sum to 0.9
Reading phenotypes for 1138 samples from ./pheno.txt
When performing cross-validation, will use 1025 samples to train models and 113 to test their accuracy
Warning, to process the data requires 6.8 Gb; if this is too high, you can reduce using "--bit-size" (currently 256)
Error, all predictors are trivial

 Interestingly, when using the Elastic Predict method alone, I encountered no issues. However, I wish to explore the potential of leveraging annotation information to improve prediction accuracy. After reviewing the LDAK documentation, I noticed that the Per-Predictor Heritabilities method is no longer recommended. I also experimented with other methods, such as SbayersRC and S-LDSC, to incorporate annotations into PRS predictions. Unfortunately, these approaches have been unsuccessful, likely due to the limited statistical power available for non-human species.
Could you kindly provide any suggestions or alternative strategies to overcome this issue? Additionally, is there a more robust method to integrate annotation data into PRS predictions in species with limited data resources?
Thank you very much for your time and assistance, and I look forward to your insights.
Best regards,
Ben
From: LDAK
Date: 2024-11-26 23:16
To: dougspeed/LDAK
CC: YangBen; Author
Subject: Re: [dougspeed/LDAK] Error, all predictors are trivial (Issue #11)
Hi, thanks for the message and for trying LDAK. Please can you provide more details (eg, the commands and screen outputs).
I think you are using --sum-hers. Looking at the LDAK code, I can not see how this error message appears.
Thanks

**Comment by dougspeed** [comment]
Hi, thanks for the extra details, they are very clear and helpful

I am still unable to find this error in LDAK, so I think I have likely removed it in my most recent versions, sorry. Really, this step should only give such an error if your correlations were computed using only trivial SNPs, but then they would have failed (otherwise, it is possible that the snps with positive per-snp heritability, happen to completely coincide with the trivial predictors, but this is very unlikely).

I will test out this command on my own data, and see I can reproduce and get back to you.

However, as a general comment, I think it will be challenging to use 9M snps with only 1000 individuals (but maybe this is ok for non-human data)

Thanks and talk soon
Doug

**Comment by dougspeed** [comment]
Hi Ben, we have just updated LDAK (linux version). Please can you try the new version, and see if it solves your problem. Thanks, Doug

**Comment by YangBen0528** [comment]
Hi 
 Thank you for the timely update. I ran the command again ：
 **~/bin/ldak6.1.linux --elastic elastic  --pheno ./pheno.txt  --bfile /public1/node3_yangb/imputation/beagel/geno --mpheno 1 --LOCO NO --ind-hers time.thin.ldak.ind.hers**
 
and it showed the following error

There are 6 pairs of arguments:
--elastic elastic
--pheno ./pheno.txt
--bfile /public1/node3_yangb/imputation/beagel/geno
--mpheno 1
--LOCO NO
--ind-hers time.thin.ldak.ind.hers

Error reading time.thin.ldak.ind.hers, Row 1 appears to have 2 columns (whereas Row 1 has 1138 columns)

This is my time.thin.ldak.ind.hers file, which is the standard output by **~/bin/ldak6.1.linux  --sum-hers time.thin.ldak --tagfile LDAK-Thin.tagging --summary time.summary --matrix LDAK-Thin.matrix --max-threads 20 --check-sums NO**
Predictor Heritability
1_7574 1.149924e-08
1_7586 1.482328e-08
1_14191 -4.620355e-08
1_14198 -1.433385e-08
1_14219 5.962980e-09
1_14245 -7.302113e-08
1_15562 1.133575e-07
1_15593 1.005364e-07
1_15612 2.119549e-06

**Comment by dougspeed** [comment]
Hi again, thanks for trying the new version so fast.
I have found the reason for the error (a bug in my code, sorry). I have edited the executable and reuploaded (it has the same name, ldak6.1.linux). If possible, please can you retry. Thanks

**Comment by YangBen0528** [comment]
Dear Doug,
﻿
I hope this message finds you well.
﻿
First and foremost, I want to express my heartfelt gratitude for the incredible work you’ve done on the new software. It has proven to be exceptionally effective and has provided us with tremendous support in our work.
﻿
Thanks to the latest version, we’ve achieved a remarkable improvement in prediction accuracy—nearly 50%—in another population. This is an outstanding achievement and a true testament to your expertise and dedication.
﻿
I also wanted to reach out for clarification regarding the heritability enrichment results. Specifically, I noticed instances where the heritability estimates are negative or accompanied by relatively large standard errors. Should these cases be considered invalid? If so, are there specific criteria or thresholds we should use to identify and exclude such results?
﻿
Here are my heritability enrichment results
Component Share SE Expected Enrichment SE Z-Stat1 Z-Stat2
Enrich_A1 -2.726033 4.662797 0.048508 -56.197150 96.123510 -0.595038 -0.672966
Enrich_A2 2.007914 6.394553 0.046121 43.535543 138.646579 0.306791 0.313785
Enrich_A3 2.015592 3.256506 0.005556 362.781686 586.130890 0.617237 0.696323
Enrich_A4 -0.424733 6.127522 0.051799 -8.199611 118.293942 -0.077769 NA
Enrich_A5 -0.366502 0.579792 0.002811 -130.370299 206.240611 -0.636976 -0.809096
Enrich_A6 0.014667 0.677586 0.003233 4.536215 209.565919 0.016874 0.016826
Enrich_A7 2.971803 2.992536 0.043792 67.861955 68.335412 0.978438 1.399243
Enrich_A8 -0.439763 1.966139 0.023836 -18.449493 82.485926 -0.235792 NA
Enrich_A9 0.027497 0.098300 0.000164 167.793082 599.845760 0.278060 NA
Enrich_A10 -3.024895 2.828262 0.024110 -125.462026 117.306384 -1.078049 -1.923717
Enrich_A11 -1.561829 2.721781 0.036467 -42.828009 74.635901 -0.587224 -0.665093
Enrich_A12 0.935804 2.435165 0.151963 6.158086 16.024675 0.321884 0.331226
Enrich_A13 -0.063064 0.171388 0.000019 -3344.045384 9088.036913 -0.368071 -0.384440
Enrich_A14 0.288127 1.492041 0.203228 1.417753 7.341719 0.056901 0.056775
Enrich_A15 -2.549913 3.650079 0.504770 -5.051631 7.231167 -0.836882 -1.044453
Enrich_A16 1.552148 0.919361 0.100000 15.521489 9.193621 1.579518 1.974549
Enrich_A17 -0.399331 0.561273 0.100000 -3.993315 5.612732 -0.889641 -1.181293
Enrich_A18 -0.547255 0.863567 0.100000 -5.472557 8.635682 -0.749513 NA
Enrich_A19 0.124271 1.036914 0.100000 1.242712 10.369145 0.023407 0.023392
Enrich_A20 -0.135617 1.665013 0.100000 -1.356174 16.650146 -0.141511 NA
Enrich_A21 -1.348847 2.420684 0.100000 -13.488484 24.206858 -0.598528 -0.667156
Enrich_A22 2.580888 3.018945 0.100000 25.808904 30.189476 0.821773 1.045077
Enrich_A23 -2.223353 2.793265 0.100000 -22.233546 27.932676 -0.831769 NA
Enrich_A24 1.926696 2.330191 0.100000 19.266977 23.301933 0.783925 0.967217
Enrich_A25 -0.529594 1.285001 0.100000 -5.295940 12.850019 -0.489956 NA
Enrich_A26 0.051343 0.199131 0.000315 162.900307 631.805988 0.256250 0.260410
Enrich_A27 -0.060148 0.144426 0.000302 -199.481349 478.989675 -0.418550 NA
Enrich_A28 -0.008021 0.061345 0.000002 -4353.049917 33294.409155 -0.130774 -0.131444
Enrich_A29 -3.139869 3.828354 0.055732 -56.338593 68.692050 -0.834719 -1.127300
Enrich_A30 1.000000 NA 1.000000 1.000000 NA NA 0.000000
Enrich_A31 0.130172 0.573685 0.000248 524.010441 2309.389107 0.226471 0.229691
Enrich_A32 -0.541816 0.943803 0.000559 -969.565294 1688.910362 -0.574670 -0.646412
Enrich_A33 -0.060440 0.284379 0.000049 -1228.298034 5779.359887 -0.212705 NA
Enrich_A34 4.627029 4.631029 0.095068 48.670600 48.712670 0.978608 NA
Enrich_A35 -0.012165 0.103587 0.000574 -21.197634 180.499664 -0.122979 NA
Enrich_A36 1.255004 2.473521 0.031275 40.127846 79.089026 0.494732 0.532576
Enrich_A37 0.022989 0.737711 0.003603 6.381047 204.762410 0.026279 NA
Enrich_Base 1.000000 NA 1.000000 1.000000 NA NA 0.000000

Thank you again for your outstanding contributions, and I greatly appreciate your guidance on this matter.

**Comment by dougspeed** [comment]
Hi, many thanks for your kind words, and I am very happy if LDAK worked.

Regarding the heritability enrichments, in general, these results are not a proble. As you say, there are some negative, but also large SD (i.e., they are not significantly negative), and so I think this just reflects that if you have many annotations, then very large sample sizes are required to get precise estimates. Further, when annotations are correlated (eg high overlap between categories), this can increase the SDs even more.

But in general, "wild" enrichment estimates are not a problem, if your aim is prediction. This is because even though the enrichment estimates are a bit unstable, prediction uses a combination of these (basically, for each predictor LDAK adds up the contributions from each annotation), and usually these are reliable, which is what matters when providing these as predictor weightings.

In terms of criteria ... if you are mainly interested in total snp heritability, and your estimate of Her_All is wild, it can help to reduce the categories, as it may be due to difficulties converging.

Thanks for your fast replies and help
Doug

**Comment by YangBen0528** [comment]
Dear Doug,
Thank you once again for your invaluable contributions. Your support has been instrumental, and I am deeply grateful.
Best regards,
Ben

**Comment by YangBen0528** [comment]
Dear Doug,

Thank you once again for your invaluable support. Your guidance has been instrumental, and I truly appreciate it.

However, I have encountered a new issue that I hope you can help clarify. When I turned off cross-validation using the following command:
**~/bin/ldak6.1.linux --elastic timeall --pheno ./pheno.txt --bfile /public1/node3_yangb/imputation/beagel/geno --mpheno 1 --LOCO NO --ind-hers time.thin.ldak.ind.hers --max-threads 80 --skip-cv YES**

The output provided effect values for 10 models and calculated prediction accuracies for another population in the range of 0.2–0.3.

However, when I enabled cross-validation with this command:
**~/bin/ldak6.1.linux --elastic elastic --pheno ./pheno.txt --bfile /public1/node3_yangb/imputation/beagel/geno --mpheno 1 --LOCO NO --ind-hers time.thin.ldak.ind.hers --max-threads 20**

The optimal model's effect values and  prediction accuracies varied significantly with each run. For instance, the first run result showed a prediction accuracy of 0.46, but subsequent runs failed to replicate this result.

From my understanding, as the article in Nature Communications, cross-validation uses only 10% of the test set to screen for the optimal model, and the final output is the estimated effect values of all individuals. I have double-checked the model parameters, which remain consistent across the 10 models in each run:
Model   p       f2
1       0.0000  1.0000
2       0.5000  0.5000
3       0.5000  0.3000
4       0.5000  0.1000
5       0.1000  0.5000
6       0.1000  0.3000
7       0.1000  0.1000
8       0.0100  0.5000
9       0.0100  0.3000
10      0.0100  0.1000
This inconsistency has left me puzzled. Could the variability be due to insufficient model convergence, causing the results to vary across runs? When I set different fixed test sets, their prediction accuracies also vary, even though their optimal models might be the same. Do you have any suggestions for ensuring more stable and reproducible results?

Thank you for your time and insights.

Best regards,
Ben

**Comment by dougspeed** [comment]
Thanks for this

Have you computed the SD of the R2 estimate (ie, the point estimate is 0.46, but do you know how precise this is?) 

Yes, you are correct, the results should be the same. E.g., suppose the default approach (internal CV) selected model 7, then when you turn off CV, and instead construct 10 PRS, the seventh of these should be identical* to that from the default approach. 

Now there are reasons for them to be different, all connected to convergence. LDAK stops when the log likelihood changes by less than n x 10e-6. This is hopefully sufficient but perhaps for some datasets, this is a poor threshold? Also, the default method will first construct PRS using the training samples, then update to allow for all samples, whereas the alternative approach will only construct using all samples ... maybe this difference affects the outcome (ie, changing the starting model).

But it would be helpful to know the SDs to tell whether the differences are significant. I compute SDs using the --jackknife function https://dougspeed.com/jackknife/

I will also test out myself, thanks

**Comment by dougspeed** [comment]
Hi again, I have done a simple test, but was unable to recreate the situation you got. Here are my commands

#make some fake data, divide into train (400 samples) and test (4000 samples)
../ldak6.1.linux --make-snps fake --num-samples 4400 --num-snps 10000 --family-size 8
../ldak6.1.linux --make-phenos fake --her .5 --num-causals 1000 --bfile fake --num-phenos 1 --power -1
shuf -n 400 fake.fam > train.fam

#do linear regression and calculate correlations
../ldak6.1.linux --linear fake --bfile fake --pheno fake.pheno --keep train.fam
../ldak6.1.linux --calc-cors fake --bfile fake --keep train.fam

#normal megaprs, with internal cv
../ldak6.1.linux --mega-prs train --cors fake --summary fake.summaries --power -1
../ldak6.1.linux --calc-scores train2 --power 0 --bfile fake --remove train.fam --scorefile train.effects --pheno fake.pheno
../ldak6.1.linux --jackknife train2 --profile train2.profile --num-blocks 200

#megaprs without cv (makes 35 models)
../ldak6.1.linux --mega-prs brain --cors fake --summary fake.summaries --power -1 --skip-cv YES
../ldak6.1.linux --calc-scores brain2 --power 0 --bfile fake --remove train.fam --scorefile brain.effects --pheno fake.pheno
../ldak6.1.linux --jackknife brain2 --profile brain2.profile --num-blocks 200

#in this case, very good agreement between normal megaprs, and skipcv versions

**Comment by YangBen0528** [comment]
Thank you for your clarification! I would like to further explain my research approach and share some challenges I am currently facing, hoping to hear your valuable insights.
﻿
In my study, I do not directly predict phenotypes. Instead, I use offspring validation to evaluate the accuracy of predictions. Specifically, I assess the correlation between the sum of parental PRS and the phenotypes of their offspring. This approach allows me to estimate the predictive accuracy of additive effects, as it focuses on the heritable component that can be passed on to offspring.
﻿
As a result, it is challenging to directly apply the method you suggested for further validation. My current hypothesis is that, although insufficient convergence may only cause small fluctuations in predictions, such fluctuations could lead to an inflation of errors when focusing on the transmissible component to offspring. This, in turn, might affect the validation results.
﻿
These are some of my thoughts and challenges. I would greatly appreciate the opportunity to discuss them with you further. Thank you very much for your time and insights!
﻿
Best regards,
Ben

**Comment by dougspeed** [comment]
Hi, thanks for sharing, that sounds a new (to me) and valid way to measure accuracy. 
However, I think you should still be able to get SDs (as per my previous reply - I think comment 9 on this thread). Specifically, if you make a list with two columns, containing the average predicted value of the parents and the (average?) phenotype of the offspring, you can then compute the correlation (as I think you have already) but also get a confdience interval by jackknifing (using the --jackknife function in LDAK) or bootstrapping (via, say, an R package)

Thanks

**Comment by YangBen0528** [comment]
Thank you very much for your help! I conducted two separate runs with different test sets. The results are as follows:
**Results from Run 1:**
Profile Measure Estimate SE
Correlation 0.365944 0.134032
Squared_correlation 0.133915 0.097072
Mean_squared_error 4.396726 0.686420
Mean_absolute_error 1.790258 0.164131
**Results from Run 2:**
Profile Measure Estimate SE
Correlation 0.297512 0.159004
Squared_correlation 0.088514 0.093071
Mean_squared_error 5.020715 0.761872
Mean_absolute_error 1.958325 0.164919
Please let me know your thoughts or if you need further clarification regarding these results. I greatly appreciate your continued support and guidance.

Best regards,
Ben

**Comment by dougspeed** [comment]
Many thanks for this

So ideally you are using independent test data (ie, the parents and offspring used in measuring accuracy are not involved when constructing the SNP weights). If that is the case, then these results indicate there is some variation in the accuracy, but it is feasible this is fine and simply reflects uncertainty (the SDs are high, relative to the accuracy)

If there is overlap between the test and training data (which may be unavoidable), then this might lead to artificially increased accuracy, and more variation across runs

Thanks

**Comment by YangBen0528** [comment]
Thank you for your clarification. I am using independent validation data in my analyses. If this results are acceptable, I will refrain from troubling you further on this matter. Once again, I greatly appreciate your time and assistance.

Best regards,
Ben
